File: sdks/python/src/echo_agent_sdk/prompt_diagnostics_values.py
from __future__ import annotations

from dataclasses import dataclass, field


@dataclass(frozen=True, slots=True)
class PromptSectionDiagnostic:
    id: str
    source: str


@dataclass(slots=True)
class PromptDiagnostics:
    sections: list[PromptSectionDiagnostic] = field(default_factory=list)

    def record(self, id: str, source: str) -> None:
        self.sections.append(PromptSectionDiagnostic(id=id, source=source))

    def count(self, id: str) -> int:
        return sum(section.id == id for section in self.sections)

File: sdks/typescript/src/prompt_diagnostics_values.ts
/** Prompt section diagnostics projected without owning prompt compilation. */
export interface PromptSectionDiagnostic {
  readonly id: string;
  readonly source: string;
}

export class PromptDiagnostics {
  public readonly sections: PromptSectionDiagnostic[];

  public constructor(sections: readonly PromptSectionDiagnostic[] = []) {
    this.sections = sections.map((section) => Object.freeze({ ...section }));
  }

  public record(id: string, source: string): void {
    this.sections.push(Object.freeze({ id, source }));
  }

  public count(id: string): number {
    return this.sections.filter((section) => section.id === id).length;
  }
}
